akka-http-caching/src/main/scala/akka/http/caching/LfuCache.scala

```diff
@@ -0,0 +1,91 @@
+package akka.http.caching
+
+import java.util.concurrent.{ CompletableFuture, Executor, TimeUnit }
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import scala.concurrent.Future
+import com.github.benmanes.caffeine.cache.{ AsyncCacheLoader, AsyncLoadingCache, Caffeine }
+import akka.http.caching.LfuCache.{ toJavaMappingFunction, dummyLoader }
+
+import scala.compat.java8.FutureConverters._
+import scala.compat.java8.FunctionConverters._
+
+object LfuCache {
+
+  /**
+   * Creates a new [[akka.http.caching.ExpiringLfuCache]] or
+   * [[akka.http.caching.SimpleLfuCache]] instance depending on whether
+   * a non-zero and finite timeToLive and/or timeToIdle is set or not.
+   */
+  def apply[V](
+    maxCapacity:     Int      = 500,
+    initialCapacity: Int      = 16,
+    timeToLive:      Duration = Duration.Inf,
+    timeToIdle:      Duration = Duration.Inf): Cache[V] = {
+
+    if (timeToLive.isFinite() || timeToIdle.isFinite())
+      new ExpiringLfuCache[V](maxCapacity, initialCapacity, timeToLive, timeToIdle)
+    else
+      new SimpleLfuCache[V](maxCapacity, initialCapacity)
+  }
+
+  //LfuCache requires a loader function on creation - this will not be used.
+  def dummyLoader[V] = new AsyncCacheLoader[Any, V] {
+    def asyncLoad(k: Any, e: Executor) =
+      Future.failed[V](new RuntimeException("Dummy loader should not be used by LfuCache")).toJava.toCompletableFuture
+  }
+
+  def toJavaMappingFunction[V](genValue: () ⇒ Future[V]) =
+    asJavaBiFunction[Any, Executor, CompletableFuture[V]]((k, e) ⇒ genValue().toJava.toCompletableFuture)
+}
+
+trait LfuCache[V] extends Cache[V] {
```

Comment:
I have a few ideas about how to improve that actual implementation code structure (but we can still do this afterwards in another PR). I'll add them in extra comments.